### PR TITLE
[Refinement] otaproxy: minor refinements and fixes 

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

> **Note**
> This PR is based on #236 and expected to be merged into main after #236 is merged.

> **Note**
> This PR superseded #225 .

> **Note**
> If possible, this PR should be included in the next releasing, but if it can not catch the releasing window it is OK to delayed the merge.

This PR fixes typing in the otaproxy module, applies refinements related to `NULL` type in orm, and refine the main entry method `retrieve_file` of `OTACache`.

### Motivation for refinement

Previously, there are some minor leftover thing not yet being done/fix for otaproxy:
1. typing is not corrected annotated in some of the modules,
2. orm.py: doesn't have a proper `NULL` type defined(just alias None type), resulting in special case handling and extra checks/code,
3. ota_cache.py: the main entry, `retrive_file` API method is too complicated and has high code cognitive complexity.

## Check list

- [x] test file(s) that covers the change(s) is implemented.
- [x] local test is passed. 
- [x] test on single VM passed.

## Changes

1. `otaproxy.orm`: implement real `NULL_TYPE` instead of type alias, `NULL_TYPE` type is singleton and returns `None` on initializing, and takes any params when init, behaviors like other data type like `int` or `str`, 
2. `otaproxy.orm`: refinement according to new `NULL_TYPE` implementaion, remove some of the special treatments for previous type alias based `NULL_TYPE` as new `NULL_TYPE` behaves like other built-in data types,
3. `otaproxy.orm`: decorate `ORMeta` with `dataclass_transfomer` to let the IDE identify `ORMBase` subclass as dataclass,
4. `otaproxy.orm`: add more inline comments,
5. `otaproxy.orm`: check field value type before field assignment, if the value type is the same as field type, do not go through type conversion,
6. `otaproxy.ota_cache`: split `_FileDescriptorHelper` into several standalone helper functions, 
7. `otaproxy.ota_cache`: refactor `retrive_file` API method, split different cases handler into standalone helper methods, improve code readability and reduce code cognitive complexity,
8. `requirement.txt`: bump `typing_extension` to 4.6.3,
9. update test files accordingly.

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] no.

## Breaking change

Does this PR introduce breaking change?
- [x] No.

### CODE REVIEW

File: otaclient/ota_proxy/utils.py [hunk 1]
```diff
@@ -39,3 +45,32 @@ async def hexdigest(self) -> str:
         return await asyncio.get_running_loop().run_in_executor(
             self._executor, self._hashf.hexdigest
         )
+
+
+class TmpCacheFileNaming:
```
Comment (Bodong-Yang) [inline otaclient/ota_proxy/utils.py:50]: New naming scheme is introduced to naming the tmp ongoing caching file predictably with the requested URL.
This makes ota_proxy aware of the relationship between a tmp file and the URL it caches for.

This is intended for the future feature of resume downloading OTA file.

File: otaclient/requirements.txt [hunk 2]
```diff
@@ -15,4 +15,4 @@ aiohttp==3.8.1
 aiofiles==22.1.0
 zstandard==0.18.0
 pycurl==7.45.1
-typing_extensions==4.5.0
\ No newline at end of file
+typing_extensions==4.6.3
```
Comment (Bodong-Yang) [inline otaclient/requirements.txt:18]: This is for the `dataclass_transformer` decorator used by `orm.ORMeta`.

File: otaclient/ota_proxy/orm.py [hunk 3]
```diff
@@ -133,6 +154,7 @@ def check_type(self, value: Any) -> bool:
         return self.type_checker(value)
 
 
+@dataclass_transform()
 class ORMeta(type):
     """This metaclass is for generating customized <TableCls>."""
```
Comment (Bodong-Yang) [inline otaclient/ota_proxy/orm.py:159]: With `dataclass_transform`, IDE now can correctly identify the `ORMBase` derived subclass as dataclass.

File: otaclient/ota_proxy/orm.py [hunk 4]
```diff
@@ -90,29 +100,40 @@ def __get__(self, obj, objtype: type) -> FV:
...
             return getattr(obj, self._private_name)  # access via instance
-        return self  # access via class, return the descriptor
+        return self  # return descriptor instance by default
 
     def __set__(self, obj, value: Any) -> None:
         if self._skipped:
-            return
-        # handle dataclass's default value setting behavior and NULL type assignment
-        if isinstance(value, type(self)) or self.field_type == NULL_TYPE:
+            return  # ignore value setting on skipped field
```
Comment (obi-t4) [inline otaclient/ota_proxy/orm.py:113]: Is this __set__ function compatible with previous version, especially `value is None`?
Or any logical fix is introduced?

Comment (Bodong-Yang) [reply otaclient/ota_proxy/orm.py:113]: The logic is the same, just the implementation is simplified. If the input `value` is None, or field type is `NULL_TYPE`, or dataclass initializing, the default value will be set.

This method is refined and simplify as `NULL_TYPE` is no-longer type alias of python's built-in `None` type. Previously in runtime `type(None)` is the same as `NULL_TYPE`, resulting we have to implement special mechanism to distinguish them when we have input value as `None`. 
This is not required anymore with this PR as `NULL_TYPE` becomes real type that different from `None` type.

Previously:
`type(None) == NULL_TYPE`, so we cannot use judgement like `self.field_type is NULL_TYPE`,
With PR:
`type(None) != NULL_TYPE`, `self.field_type is NULL_TYPE` can be used now.

**Review by obi-t4 — APPROVED** [review]
:+1:

**Pull request merged**